codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java

```diff
@@ -510,31 +499,15 @@ public void onUnknownFrame(ChannelHandlerContext ctx, byte frameType, int stream
             onUnknownFrame0(ctx, frameType, streamId, flags, payload);
         }
 
-        /**
-         * Indicates whether or not frames for the given stream should be ignored based on the state of the
-         * stream/connection.
-         */
-        private boolean shouldIgnoreFrame(Http2Stream stream, boolean allowResetSent) {
-            if (connection.goAwaySent() &&
-                    (stream == null || connection.remote().lastStreamCreated() <= stream.id())) {
-                // Frames from streams created after we sent a go-away should be ignored.
-                // Frames for the connection stream ID (i.e. 0) will always be allowed.
-                return true;
-            }
-
-            // Also ignore inbound frames after we sent a RST_STREAM frame.
-            return stream != null && !allowResetSent && stream.isResetSent();
+        private boolean streamCreatedAfterGoAwaySent(Http2Stream stream) {
+            int lastStreamId = connection().remote().lastKnownStream();
+            // Ignore inbound frames after a GOAWAY was sent and either the stream does not exist (anymore) or
+            // the streams id is greater than the last stream id set in the GOAWAY frame.
+            return connection.goAwaySent() && (stream == null || stream.id() > lastStreamId);
         }
 
-        /**
-         * Verifies that a GO_AWAY frame was not previously received from the remote endpoint. If it was, throws a
-         * connection error.
-         */
-        private void verifyGoAwayNotReceived() throws Http2Exception {
-            if (connection.goAwayReceived()) {
-                // Connection error.
-                throw connectionError(PROTOCOL_ERROR, "Received frames after receiving GO_AWAY");
-            }
+        private boolean resetSentForStream(Http2Stream stream) {
+            return stream != null && stream.isResetSent();
```

Comment:
If we are no longer guaranteed that we will maintain stream objects for 5 seconds, should we separately maintain some history of streams that have recently closed and under what conditions?  This history could then be garbage collected in a similar way to what we were doing before.  Something like this perhaps:

```java
class StreamHistory {
   enum CloseState {
       CLOSED_NORMALLY,
       RESET_SENT,
       RESET_RECEIVED
   }
   int id;
   long time;
   CloseState closeState;
}
```

This should be much lighter weight than keeping the entire stream around, and it might help us to make better decisions about what to do about historical streams.  WDYT?